library/x509write_crt.c

```diff
@@ -504,9 +531,18 @@ int mbedtls_x509write_crt_der( mbedtls_x509write_cert *ctx,
 
     /*
      *  Serial   ::=  INTEGER
+     *
+     * Written data is:
+     * - [ctx->serial_len] bytes for the raw serial buffer
+     * - 1 byte for the length
+     * - 1 byte for the TAG
      */
-    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_mpi( &c, buf,
-                                                       &ctx->serial ) );
+    MBEDTLS_ASN1_CHK_ADD( len, mbedtls_asn1_write_raw_buffer( &c, buf,
+                                ctx->serial, ctx->serial_len) );
```

Comment:
I should have noticed yesterday, but I don't think that's correct: when the most significant bit of the first byte is set, we need to prepend a 00 byte. Otherwise when decoding the leading 1 bit will be interpreted as a sign bit.

See [what we do](https://github.com/Mbed-TLS/mbedtls/blob/development/library/asn1write.c#L144) in `mbedtls_asn1_write_mpi()` - except here we don't need the `X->s == 1` part of the tests - we know we only deal with non-negative numbers.

PS: I tend to use [A Layman's Guide to ASN.1](https://luca.ntop.org/Teaching/Appunti/asn1.html) as my go-to reference as that's the first place where I learned about ASN.1 and its encodings. It has as few examples under "5.7 INTEGER".

Reply:
If I remember correctly, `mbedtls_asn1_write_mpi` is the function I looked before doing this change, but apparently I totally missed one very important step :facepalm: 

Btw thanks for the guide!